File: pythonlibrary/aibrd/extractors/module_detector.py
from ..llm.client import call_llm_json
from ..models.module import DetectedModule
from ..registry import derive_prefix

_DETECT_SYSTEM = """You are a business analyst. Detect major business domains/modules in this BRD.
Rules:
- A module is a distinct business capability (e.g. "Payment Processing", "User Authentication")
- Ignore generic sections: Introduction, Glossary, Appendix, Overview
- Use short display names (2-3 words)
- Slugs must be lowercase, hyphen-separated
- Return 1 module for small/single-domain BRDs
Return JSON: {"modules": [{"display_name": "...", "slug": "...", "confidence": "high|low"}]}"""

_MATCH_SYSTEM = """You are a business analyst. Given a list of existing modules and a new requirement,
determine which module it belongs to, or suggest a new one if none fit.
Return JSON: {"slug": "existing-or-new-slug", "display_name": "...", "is_new": true|false, "confidence": "high|low"}"""


def detect_modules(brd_text: str) -> list[DetectedModule]:
    raw = call_llm_json(brd_text[:12000], _DETECT_SYSTEM)
    prefixes_seen: list[str] = []
    result = []
    for m in raw.get("modules", []):
        prefix = derive_prefix(m["slug"], prefixes_seen)
        prefixes_seen.append(prefix)
        result.append(DetectedModule(
            display_name=m["display_name"],
            slug=m["slug"],
            prefix=prefix,
            confidence=m.get("confidence", "high"),
        ))
    return result


def match_or_create_module(
    requirement_text: str,
    existing: list[DetectedModule],
) -> DetectedModule:
    if not existing:
        detected = detect_modules(requirement_text)
        return detected[0] if detected else DetectedModule(
            display_name="General", slug="general", prefix="GEN", confidence="low"
        )

    module_list = "\n".join(f"{m.slug}: {m.display_name}" for m in existing)
    prompt = f"Existing modules:\n{module_list}\n\nNew requirement:\n{requirement_text}"
    result = call_llm_json(prompt, _MATCH_SYSTEM)

    if not result.get("is_new"):
        matched = next((m for m in existing if m.slug == result["slug"]), None)
        if matched:
            return matched

    existing_prefixes = [m.prefix for m in existing]
    prefix = derive_prefix(result["slug"], existing_prefixes)
    return DetectedModule(
        display_name=result["display_name"],
        slug=result["slug"],
        prefix=prefix,
        confidence=result.get("confidence", "low"),
    )

File: vscode-extension/src/core/extractors/module_detector.ts
import * as vscode from 'vscode'
import { callLLMJson } from '../../llm/client'
import { DetectedModule } from '../models/module'

interface RawDetected {
  modules: Array<{
    displayName: string
    slug: string
    confidence: 'high' | 'low'
  }>
}

const SYSTEM = `You are a business analyst. Detect the major business domains/modules in this BRD text.
Rules:
- A module is a distinct business capability (e.g. "Payment Processing", "User Authentication")
- Ignore generic sections like "Introduction", "Glossary", "Appendix"
- Use short, clear display names (2-3 words max)
- Slugs must be lowercase, hyphen-separated, no spaces
- Return 1 module for small/single-domain BRDs`

export async function detectModules(
  brdText: string,
  token?: vscode.CancellationToken
): Promise<DetectedModule[]> {
  const raw = await callLLMJson<RawDetected>(
    brdText.slice(0, 12000),
    SYSTEM,
    token
  )

  const prefixesSeen: string[] = []
  return raw.modules.map(m => {
    const prefix = derivePrefix(m.slug, prefixesSeen)
    prefixesSeen.push(prefix)
    return { displayName: m.displayName, slug: m.slug, prefix, confidence: m.confidence }
  })
}

export async function matchOrCreateModule(
  requirementText: string,
  existingModules: DetectedModule[],
  token?: vscode.CancellationToken
): Promise<DetectedModule> {
  if (existingModules.length === 0) {
    const detected = await detectModules(requirementText, token)
    return detected[0] ?? { displayName: 'General', slug: 'general', prefix: 'GEN', confidence: 'low' }
  }

  const moduleList = existingModules.map(m => `${m.slug}: ${m.displayName}`).join('\n')
  const prompt = `Existing modules:\n${moduleList}\n\nNew requirement:\n${requirementText}\n\nWhich existing module does this belong to? If none fit, suggest a new one.`
  const MATCH_SYSTEM = `You are a business analyst. Return JSON: { "slug": "existing-slug-or-new", "displayName": "...", "isNew": true|false, "confidence": "high"|"low" }`

  interface MatchResult { slug: string; displayName: string; isNew: boolean; confidence: 'high' | 'low' }
  const result = await callLLMJson<MatchResult>(prompt, MATCH_SYSTEM, token)

  if (!result.isNew) {
    return existingModules.find(m => m.slug === result.slug) ?? existingModules[0]
  }

  const prefixesSeen = existingModules.map(m => m.prefix)
  const prefix = derivePrefix(result.slug, prefixesSeen)
  return { displayName: result.displayName, slug: result.slug, prefix, confidence: result.confidence }
}

function derivePrefix(slug: string, existing: string[]): string {
  const words = slug.replace(/-/g, ' ').split(' ').filter(Boolean)
  const candidates = [
    words[0].slice(0, 3).toUpperCase(),
    words.map(w => w[0]).join('').toUpperCase(),
    words[0].slice(0, 4).toUpperCase()
  ]
  for (const c of candidates) {
    if (!existing.includes(c)) return c
  }
  let n = 2
  while (existing.includes(candidates[0] + n)) n++
  return candidates[0] + n
}
